functionary/prompt_template/base_template.py

```diff
@@ -147,25 +156,30 @@ def update_grammar_sampling_gen_state(
             # Either '{' or '{"' or '{}'
             elif self.fn_param_sep_token in gen_state["curr_text"]:
                 # Check if no arguments are called and go straight to "pre-function"
-                if gen_state["curr_text"].endswith("}"):
+                if "}" in gen_state["curr_text"]:
                     gen_state["stage"] = "pre-function"
-                elif gen_state["curr_text"].endswith('"'):
+                elif '"' in gen_state["curr_text"]:
                     gen_state["stage"] = "parameter-name"
-                    gen_state["curr_text"], gen_state["curr_tokens"] = "", []
+                    if gen_state["curr_text"].endswith('"'):
+                        gen_state["curr_text"], gen_state["curr_tokens"] = "", []
+                    else:
+                        gen_state["curr_tokens"] = [new_token_id]
+                        gen_state["curr_text"] = tokenizer.decode([new_token_id])
         elif gen_state["stage"] == "parameter-name":
-            # Check if no arguments are called and go straight to "pre-function"
-            if gen_state["curr_text"] == self.fn_param_sep_token + "}":
-                gen_state["stage"] = "pre-function"
-            else:
-                # Get the latest param
-                latest_param_str = gen_state["curr_text"]
+            # Get the latest param
+            latest_param_str = gen_state["curr_text"]
 
-                # Check if the new state is in "parameter-value" stage
-                stop_token = self.get_stop_token_for_function_parameter(
-                    stage="parameter"
-                )
+            # Remove unneccesary prefixes before the parameter-name part
+            if len(gen_state["curr_tokens"]) > 0 and '"' in tokenizer.decode(
+                [gen_state["curr_tokens"][0]]
+            ):
+                latest_param_str = latest_param_str[latest_param_str.find('"') + 1 :]
+
+            # Check if the new state is in "parameter-value" stage
+            stop_token = self.get_stop_token_for_function_parameter(stage="parameter")
+            if stop_token in latest_param_str:
                 pattern = stop_token + r".*$"
-                match_res = re.search(pattern, latest_param_str)
+                match_res = re.search(pattern, latest_param_str, re.DOTALL)
```

Comment:
This one is not popular but: stop_token='":' is not general enough; can also be '"   :' ? right ?
for example:
{"a"   : 10}

Reply:
For this part, it is ok because a parameter name will always be string. So it will always be `"parameter-name": {parameter-value}` ONLY

Reply:
Once `"parameter-name": ` is detected, we will go to parameter-value stage already.